### PR TITLE
Stories/outbound link hoc

### PR DESCRIPTION
**Pull request by Xyfi**
## Summary

This PR can be summarized in the following changelog entry:

* Adds a decorator function to create an outbound button with a11y functionality.

## Test instructions

This PR can be tested by following these steps:
ContentAnalysis.js:
```
import React from "react";
import styled from "styled-components";
import { IntlProvider } from "react-intl";

import { makeOutboundLink } from "../../../../utils/makeOutboundLink";
import { BaseButton, Button, IconButton } from "../../Shared/components/Button";
import { YoastButton } from "../../Shared/components/YoastButton";
import { edit } from "../../../../style-guide/svg";
import { BaseLinkButton, LinkButton } from "../../Shared/components/LinkButton";
import { YoastLinkButton } from "../../Shared/components/YoastLinkButton";

const StyledAnchor = styled.a`
	display: inline-block;
	padding: 8px;
	border: 1px solid black;
	border-radius: 8px;
	text-decoration: none;
	color: black;
	:hover {
		background: #DDD;
	}
`;

const OutboundYoastButton = makeOutboundLink( YoastLinkButton );
const StyledOutboundAnchor = makeOutboundLink( StyledAnchor );
const OutboundAnchor = makeOutboundLink();

export const ContentAnalysisContainer = styled.div`
	min-height: 700px;
	width: 100%;
	padding: 40px;
	background-color: white;

	a,
	button {
		margin-right: 16px;
	}

	.no-margin {
		margin: 0;
	}

	.taller {
		min-height: 152px;
	}

	.with-max-width {
		max-width: 152px;
	}

	.test-large-button {
		min-width: 200px;
	}
`;

/**
 * Returns the ContentAnalysis component.
 *
 * @returns {ReactElement} The ContentAnalysis component.
 */
export default function ContentAnalysis() {
	return <IntlProvider>
		<ContentAnalysisContainer>
			<h2>Outbound links</h2>
			<OutboundYoastButton href="https://www.yoast.com">Go to Yoast.com</OutboundYoastButton>
			<OutboundYoastButton href="https://www.yoast.com" className="taller">Go to Yoast.com</OutboundYoastButton>
			<br />
			<StyledOutboundAnchor href="#">Styled anchor</StyledOutboundAnchor>
			<br /><br />
			Some text <OutboundAnchor href="#" className="no-margin">Normal anchor</OutboundAnchor> and some more text.

			<h2>Buttons</h2>
			<BaseButton>Base Button</BaseButton>
			<Button>Button</Button>
			<IconButton icon={ edit } iconColor="#c00" aria-label="IconButton with icon only" />
			<IconButton icon={ edit } iconColor="#c00">Icon Button</IconButton>
			<IconButton icon={ edit } iconColor="#c00" className="with-max-width">Icon Button a very long text here bla bla</IconButton>
			<YoastButton onClick={ () => {
				console.log( "hello Button" );
			} }>Yoast Button</YoastButton>
			<YoastButton backgroundColor="lightblue" textColor="#333" withTextShadow={ false }>Yoast Button</YoastButton>
			<YoastButton className="test-large-button">Test Button</YoastButton>

			<h2>Links</h2>
			<BaseLinkButton href="#somewhere1">Base Button</BaseLinkButton>
			<LinkButton href="#somewhere2">Button</LinkButton>
			<YoastLinkButton href="#somewhere3">Yoast Button</YoastLinkButton>
			<YoastLinkButton href="#somewhere4" backgroundColor="lightblue" textColor="#333" withTextShadow={ false }>Yoast Button</YoastLinkButton>
			<YoastLinkButton className="test-large-button" href="#somewhere3">Test Button</YoastLinkButton>
		</ContentAnalysisContainer>
	</IntlProvider>;
}

```

### CODE REVIEW

File: utils/makeOutboundLink.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { defineMessages, injectIntl, intlShape } from "react-intl";
+
+import { A11yNotice } from "../composites/Plugin/Shared/components/A11yNotice";
+
+const messages = defineMessages( {
+	opensInNewTab: {
+		id: "a11yNotice.opensInNewTab",
+		defaultMessage: "(This link opens in a new tab)",
```
Comment (IreneStr) [inline utils/makeOutboundLink.js]: Please remove brackets

File: utils/makeOutboundLink.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	},
+} );
+
+/**
+ * Makes an anchor component into an outbound link that opens in a new tab.
+ *
+ * @param {ReactElement} Component The anchor component to be wrapped.
+ *
+ * @returns {OutboundLink} The OutboundLink component.
```
Comment (IreneStr) [inline utils/makeOutboundLink.js]: ReactElement?

File: utils/makeOutboundLink.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	class OutboundLink extends React.Component {
+		render() {
+			const newProps = Object.assign(
+				{
+					target: "_blank",
+					rel: "noopener noreferrer",
+				},
+				this.props
+			);
+			// Using React.createElement because it can accept a string as a component parameter
```
Comment (IreneStr) [inline utils/makeOutboundLink.js]: Use

**Comment by IreneStr** [comment]
Additional issues:

- 'This link opens in a new tab' is announced before the anchor text.
schermafbeelding 2017-09-27 om 09 54 38

- (In Safari, not in Chrome:) The styling of the span is off.

schermafbeelding 2017-09-27 om 09 49 21

**Comment by afercia** [comment]
Please test with the updated code posted above.

Tries to fix the Safari + VoiceOver bug about the position of `screen-reader-text` using:
`transform: translateY(1em);`

Hopefully this should work under all conditions. The visually hidden text gets just shifted down a bit, depending on the inherited font-size.  I've added all buttons and links to test in depth and check this doesn't break anything.

Also, moved a CSS rule meant to fix a Safari 10 flexbox bugs that happens only for buttons, in a place where it applies just to buttons and not links.
Worth noting Safari is now 11 and seems to have fixed this bug. To reproduce it you should test in Safari 10.

Also, please keep the parenthesis around the `Opens in a new browser tab` text. Parenthesis make screen readers pause briefly after the link text, and thats a good thing. Also consistent with same text used in My Yoast.

**Comment by afercia** [comment]
For clarity: when checking the position of the `screen-reader-text` don't let the Chrome inspector fool you. Sometimes it will indicate the position slightly on the right of the actual position. See the screenshot below, there the actual position is the highlighted single pixel:

screen shot 2017-09-27 at 12 33 31

**Comment by IreneStr** [comment]
CR and acceptance done 👍 
Tested with VoiceOver in Safari 10 and 11

**Pull request merged**